README.md
# eclipxe/buzoncfdi-cfdireader

**Library to read and validate a Mexican CFDI 3.2 and 3.3 (Comprobantre Fiscal por Internet)**

[![Gitter][badge-gitter]][gitter]
[![Latest Version][badge-release]][release]
[![Software License][badge-license]][license]
[![Build Status][badge-build]][build]
[![Scrutinizer][badge-quality]][quality]
[![Coverage Status][badge-coverage]][coverage]
[![Total Downloads][badge-downloads]][downloads]
[![SensioLabsInsight][badge-sensiolabs]][sensiolabs]

This library open an Xml CFDI and read it as a SimpleXML (without namespaces) for easy access.

It also validates the CFDI agains it's XSD files
using [Xml Schema Validator](https://github.com/eclipxe13/XmlSchemaValidator) library.

This library is part of buzoncfdi project, be aware that this could change since the hole project is on development.

## Install

Install using composer like `composer require eclipxe/buzoncfdi-cfdireader`

## Basic usage

```php
<?php
// get the contents from a file or whatever your source is
$xml = file_get_contents('some-cfdi-example.xml');

// create the reader
$reader = new \CFDIReader\CFDIReader($xml);

// The root element is retrieved by comprobante function, it returns always a new instance (cloned) of the root element
/** @var \SimpleXMLElement $cfdi */
$cfdi = $reader->comprobante();

// all the nodes and attributes first letter is in lower case except if the attribute is all upper case
echo $cfdi->complemento->timbreFiscalDigital["UUID"];
```

### scripts/validate.php

Use `php scripts/validate.php [file1.xml] [file2.xml]` to test CFDIs and see the results.

Options:
- `--local-path | -l local-path`: use the `local-path` argument as the path of local storage.
    - If `local-path` is the text `disable` then no local storage will be used.
    - If `local-path` is not set or is an empty text then will use the location of the library plus `/resources`

## Create a reader

The `CFDIReader` class is immutable, it only perform the following checks:

* The content must be a valid XML
* The root element must be Comprobante
* The version attribute must be 3.2 or 3.3
* The namespaces must include http://www.sat.gob.mx/cfd/3
* If set, the namespaces must include http://www.sat.gob.mx/TimbreFiscalDigital
  and the element Comprobante/Complemento/TimbreFiscalDigital must exists

### CFDIReader helper functions

`\CFDIReader\CFDIReader::node`: Help you to retrieve an specific node, returns `NULL` if not found.
i. e. `$cfdi->node('conceptos', 'concepto')` will return the list of `Complemento/Conceptos/Concepto` nodes.

`\CFDIReader\CFDIReader::attribute`: Help you to retrieve an specific attribute inside a node,
returns an empty string if the node or the attribute was not found.
i. e. `$cfdi->node('emisor', 'nombre')` will return something like `'Empresa de ejemplo SA de CV'`.

## Using the factory

The `CFDIFactory` helps to create `CFDIReaders` validating against `SchemasValidator` and `PostValidator`.

Validate a CFDI version 3.3 against its schemas (XSD Files) will require (as of 2017-09-01) more than 6.3 MB.
This means that you should keep local copies of all xsd files.

This is why this library depends on `eclipxe/xmlresourceretriever` and uses `CFDIFactory` to help on this
setup.

If you set the property `CFDIFactory::localResourcesPath` with:
- `NULL`: it will take the installation directory of the library and append `/resources` as the local repository.
- non empty string: It will take the argument as the local repository path.
- empty string: disable local repository and use internet files.

The PostValidator do some specific checks about the CFDI, this includes `Conceptos`, `Fechas`, `Impuestos` and `Totales`.

## About Addendas and XML Validation

> _why don't you create valid XML files!?_

An XML file has a strict specification, if it includes XML Schemas then the specification must be followed.

The CFDI spec say that it is valid to include additional nodes inside the Addenda but it must follow the
XML specification (including namespaces and schemas).

The problem is that, since the addenda is not part of the cadena de origen, and therefore not part of the
seal (sello); the emmiters can include additional nodes inside the Addenda after it was signed without
really breaking the CFDI but breaking the XML validation.

So, can I edit a CFDI? Yes. As long you don't change any content of the source string (cadena de origen).

I had created an utility named `CFDICleaner` that removes Addendas and unused namespaces declarations.
You can use this tool to validate the document without this garbage.

```php
<?php
// $content as a clean version of the
$content = \CFDIReader\CFDICleaner::staticClean(file_get_contents('cfdi-dirty.xml'));
```

## Contributing

There is a lot of work, this is an open source project that try to offer a framework agnostic way to deal with
Mexican CFDI version 3.2 and 3.3.

Contributions are welcome! Please read [CONTRIBUTING][] for details
and don't forget to take a look in the [TODO][] and [CHANGELOG][] files.

## License

The eclipxe/buzoncfdi-cfdireader library is copyright © [Carlos C Soto](https://eclipxe.com.mx/)
and licensed for use under the MIT License (MIT). Please see [LICENSE][] for more information.

[contributing]: https://github.com/eclipxe13/buzoncfdi-cfdireader/blob/master/CONTRIBUTING.md
[changelog]: https://github.com/eclipxe13/buzoncfdi-cfdireader/blob/master/CHANGELOG.md
[todo]: https://github.com/eclipxe13/buzoncfdi-cfdireader/blob/master/TODO.md

[release]: https://github.com/eclipxe13/buzoncfdi-cfdireader/releases
[license]: https://github.com/eclipxe13/buzoncfdi-cfdireader/blob/master/LICENSE
[gitter]: https://gitter.im/eclipxe13/buzoncfdi-cfdireader?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge
[build]: https://travis-ci.org/eclipxe13/buzoncfdi-cfdireader?branch=master
[quality]: https://scrutinizer-ci.com/g/eclipxe13/buzoncfdi-cfdireader/
[coverage]: https://scrutinizer-ci.com/g/eclipxe13/buzoncfdi-cfdireader/code-structure/master/code-coverage
[downloads]: https://packagist.org/packages/eclipxe/buzoncfdi-cfdireader
[sensiolabs]: https://insight.sensiolabs.com/projects/ffa9eb49-58e3-4532-acdd-f8089d46ad73

[badge-gitter]: https://badges.gitter.im/eclipxe13/buzoncfdi-cfdireader.svg
[badge-release]: https://img.shields.io/github/tag/eclipxe13/buzoncfdi-cfdireader.svg?label=version&style=flat-square
[badge-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[badge-build]: https://img.shields.io/travis/eclipxe13/buzoncfdi-cfdireader/master.svg?style=flat-square
[badge-quality]: https://img.shields.io/scrutinizer/g/eclipxe13/buzoncfdi-cfdireader/master.svg?style=flat-square
[badge-coverage]: https://img.shields.io/scrutinizer/coverage/g/eclipxe13/buzoncfdi-cfdireader/master.svg?style=flat-square
[badge-downloads]: https://img.shields.io/packagist/dt/eclipxe/buzoncfdi-cfdireader.svg?style=flat-square
[badge-sensiolabs]: https://insight.sensiolabs.com/projects/ffa9eb49-58e3-4532-acdd-f8089d46ad73/mini.png


CHANGELOG.md
This project is using Semantic Versioning, meaning that a version is: MAJOR.MINOR.PATCH
  - MAJOR version when you make incompatible API changes,
  - MINOR version when you add functionality in a backwards-compatible manner, and
  - PATCH version when you make backwards-compatible bug fixes.

# Version 2.5.2
- Fix: Fechas validator take a tolerance of 60 seconds to avoid time sync errors.
  The property can be adjusted.

# Version 2.5.1
- Fix: `PostValidator` was using local issues property instead of variable.
- Fix: Probable bug when comparing null to string (phpstan complains).
- Typos and minor changes

# Version 2.5.0 - CFDI and TFD version correlation
- Add `TFDVersions` to validate correlation between CFDI and TFD versions
    - CFDI version 3.3 uses TFD version 1.1
    - CFDI version 3.2 uses TFD version 1.0
  This was set according to http://www.sat.gob.mx/informacion_fiscal/factura_electronica/Paginas/timbre_fiscal.aspx and a chat session with authorities (SAT).
- Add `TFDVersions` to `CFDIFactory::newPostValidator`


# Version 2.4.0
- Fix bug on initialize `CFDIReader\Scripts\Validate` local path
- Deprecate protected property $comprobante inside `\CFDIReader\PostValidations\Validators\AbstractValidator`
- Stop using previously deprecated property in internal validators
- Docblock `\CFDIReader\PostValidations\Validators\AbstractValidator` as access private
- Fix project description


# Version 2.3.0 - node and attribute
- Add `node` and `attribute` methods to `CFDIReader`.
  This action simplify the information extraction from the `Comprobante`node.
  - `nodes` returns `NULL` or a cloned `\SimpleXmlElement`
  - `attribute` returns a string
- Refactor validators and `CFDIReader` to use this helpers, now is more readable and simple
- Fix possible bug when `CFDIFactory::newXsltRetriever()` returns `NULL`
- Fix possible bug inside `CFDICleaner::xpathQuery()` when `\DOMXPath::query` returns `FALSE`
- Fix docblocks of object type properties that allows nulls
- Special thanks to `phpstan/phpstan` to help me catch possible bugs

# Version 2.2.0 - Validate certificate
- Create a new post validator `\CFDIReader\PostValidations\Validators\Certificado` that checks:
    - certificate number match (error)
    - emisor rfc match  (error)
    - emisor nombre match  (warning)
    - date between certificate dates (error)
    - if contains a CadenaOrigen object, verify that the sello match with the "cadena de origen" using
      the public key certificate (error). If this tails then the CFDI was modified.
- Add dependency on `eclipxe/CfdiUtils` to be able to create the "Cadena Origen" and recover
  a certificate from a cfdi.
- Add new methods on `CFDIReader\CFDIFactory` to create helper objects:
    - `newXsltRetriever(DownloaderInterface $downloader = null): ?XsltRetriever`
      Create a new instance of `XmlResourceRetriever\XsltRetriever` based on local resources path.
      This allow to create a local repository of xslt files.
      This is very similar to `newRetriever(DownloaderInterface $downloader = null): ?XsdRetriever`
      but for XSLT resources instead of XSD.
    - `newCadenaOrigen(): CadenaOrigen`
      Create a new instance of `CfdiUtils\CadenaOrigen`, using `newXsltRetriever`
    - `newCertificadoValidator(): CertificadoValidator`
      Create a new instance of `CFDIReader\PostValidations\Validators\Certificado` setting the
      `CadenaOrigen` with the returned instance of `newCadenaOrigen`

# Version 2.1.0 
- Add `source(): string` and `document(): \DOMDocument` methods to `CFDIReader\CFDIReader`
- Add docblocks to `CFDIReader\PostValidations\IssuesTypes` constants
- Add docblocks to `CFDIReader\PostValidations\ValidatorInterface` methods
- Improve test files:
    - Remove unmeaning asserts
    - By default setup the test does not require the cfdi with timbre
    - Add asset/v32/real.xml with a real cfdi
    - Use correct Emisor/nombre (from certificate)
    - Change v33/valid.xml to include: Emisor, Receptor, Fecha, Sello, Certificado, NoCertificado, FechaTimbrado
    - Update v33/valid-without-timbre.xml according to valid
- Update TODO.md

# Version 2.0.0
- Require PHP 7.0
- Add scalar type declarations and docblocks @return void
- Update dependence XmlSchemaValidator.
    - That class no longer retrieve and store a local copy of resources,
      implement `eclipxe/xmlresourceretriever` library
- Remove SchemaValidator class if favor of `CFDIReader\SchemasValidator\SchemasValidator`
- Add `CFDIReader::getVersion()`
- Remove PHP 5.6 from build matrix

# Version 1.2.0
- Make optional to require that the Comprobante contains the TimbreFiscalDigital.
  This is useful to validate before obtaining the TimbreFiscalDigital from an
  authorized third party.

# Version 1.1.1
- Improved code coverage, thanks to pull request #5 by @driftking301
- Sort assets by cfdi version
- Won't publish on packagist, just update master branch

# Version 1.1.0
- Fix `CFDIFactory::newLocator` method since it was not registering in the correct way the commonxsd files
- Add **basic support for CFDI 3.3**
  There are several thinks that can be made since version 3.3
  is not only about XML/XSD validations but also "logic" validations.
- Add `CFDIReader::allowedVersions` method that return an array with 3.2 and 3.3
- Add support for CFDI 3.3 in `CFDICleaner`
- Fix isolation to use libxml internal errors in `CFDICleaner::loadContent`
- Include local versions for `cfdv33.xsd` and `TimbreFiscalDigitalv11.xsd` modified to point to local files
  of `catCFDI.xsd` and `tdCFDI.xsd`  
- Fix documentation, include gitter chat room in README
- Fix composer.json required versions (use ^X.Y instead of @stable)
- Add doc/ files (in spanish) about the project
- Add build on scrutinizer over PHP versions 5.6, 7.0 and 7.1

# Version 1.0.11
- Add docblocks to CFDIReader\Scripts\Validate class
- Improve code coverage for CFDIReader\Scripts\Validate
- Reduce complexity of CFDIReader\Scripts\Validate::run

src/CFDIReader/CFDIReader.php
<?php
namespace CFDIReader;

use DOMDocument;
use SimpleXMLElement;

/**
 * CFDI Reader immutable class to recover contents from a CFDI.
 * This task is a kind of difficult since a CFDI can contain
 * several namespaces and include different rules than the need by SAT.
 *
 * The only mandatory namespaces is:
 * http://www.sat.gob.mx/cfd/3 for CFDI v3.2 and v3.3
 *
 * By default it makes also this namespace as mandatory but can be ommited by constructor:
 * http://www.sat.gob.mx/TimbreFiscalDigital for TimbreFiscalDigital (Seal) versions 1.0 and 1.1
 *
 * The class do not perform validations, only very basic as:
 * - Content must be a XML string
 * - Content must implement mandatory namespaces
 * - Root node must be Comprobante
 * - Root node must contain an attribute version with the value 3.2 or 3.3
 * - The node Comprobante/Complemento/TimbreFiscalDigital must exists if set in the constructor
 *
 * Other validations like XSD can be made using SchemaValidator
 * To validate the logic of the content you can use PostValidations helpers
 *
 * @package CFDIReader
 */
class CFDIReader
{
    /** @var SimpleXMLElement */
    private $comprobante;

    /** @var string */
    private $source;

    /**
     * Return an array of the versions that the reader can process
     *
     * @return string[]
     */
    public static function allowedVersions(): array
    {
        return ['3.2', '3.3'];
    }

    /**
     * @see CFDIReader
     * @param string $content xml contents
     * @param bool $requireTimbre
     * @throws \InvalidArgumentException when the content is not a valid XML
     */
    public function __construct(string $content, bool $requireTimbre = true)
    {
        // create the SimpleXMLElement
        try {
            $xml = new SimpleXMLElement($content);
        } catch (\Exception $ex) {
            throw new \InvalidArgumentException(
                'The content provided to build the CFDIReader is not a valid XML',
                null,
                $ex
            );
        }
        // check the root node name
        if ('Comprobante' !== $xml->getName()) {
            throw new \InvalidArgumentException('The XML root node must be Comprobante');
        }
        $version = '';
        if (isset($xml['version'])) {
            $version = strval($xml['version']);
        } elseif (isset($xml['Version'])) {
            $version = strval($xml['Version']);
        }
        if (! in_array($version, $this->allowedVersions())) {
            throw new \InvalidArgumentException(
                sprintf(
                    'The Comprobante version must be one of the following: %s.',
                    implode(', ', $this->allowedVersions())
                )
            );
        }
        // check it contains both mandatory namespaces
        $nss = array_values($xml->getNamespaces(true));
        $required = ['http://www.sat.gob.mx/cfd/3'];
        if ($requireTimbre) {
            $required[] = 'http://www.sat.gob.mx/TimbreFiscalDigital';
        }
        foreach ($required as $namespace) {
            if (! in_array($namespace, $nss)) {
                throw new \InvalidArgumentException('The content does not use the namespace ' . $namespace);
            }
        }
        // include a null element to also copy the elements without namespace
        array_push($nss, null);
        // populate the root element
        $dummy = new SimpleXMLElement('<dummy/>');
        $this->comprobante = $this->appendChild($xml, $dummy, $nss);
        // check that it contains the node comprobante/complemento/timbreFiscalDigital if required
        if ($requireTimbre && ! $this->hasTimbreFiscalDigital()) {
            throw new \InvalidArgumentException('Seal not found on Comprobante/Complemento/TimbreFiscalDigital');
        }
        $this->source = $content;
    }

    /**
     * Get the xml source that was used to create this object
     *
     * @return string
     */
    public function source(): string
    {
        return $this->source;
    }

    /**
     * Retrieve a new instance of a DOMDocument using source as the content.
     * It always creates and return a new instance
     *
     * @return DOMDocument
     */
    public function document(): DOMDocument
    {
        $document = new DOMDocument();
        $document->loadXML($this->source());
        return $document;
    }

    /**
     * Get a copy of the root element
     *
     * @return SimpleXMLElement
     */
    public function comprobante()
    {
        return clone $this->comprobante;
    }

    public function getVersion(): string
    {
        return $this->attribute('version');
    }

    /**
     * Return the node in the path inside the Comprobante
     * Returns null if the node does not exists
     * The node retrieved is always a copy
     *
     * @param string[] ...$nodePath
     * @return SimpleXMLElement|null
     */
    public function node(...$nodePath)
    {
        $node = $this->retrieveNode(...$nodePath);
        return (null === $node) ? null : clone $node;
    }

    /**
     * Get the attribute content of a comprobante or child node
     * Return an empty string if the node or the attribute does not exists
     *
     * The last argument is always the attribute name
     *
     * @param string[] ...$nodePath
     * @return string
     */
    public function attribute(string ...$nodePath): string
    {
        // cast to string since array_pop can return NULL
        $attribute = (string) array_pop($nodePath);
        $node = $this->retrieveNode(...$nodePath);
        return (null !== $node) ? (string) $node[$attribute] : '';
    }

    /**
     * Get the UUID from the document. If the node does not exists then return an empty string
     *
     * @return string
     */
    public function getUUID(): string
    {
        return $this->attribute('complemento', 'timbreFiscalDigital', 'UUID');
    }

    /**
     * Return true if the node Comprobante/Complemento/TimbreFiscalDigital exists
     *
     * @return bool
     */
    public function hasTimbreFiscalDigital(): bool
    {
        return (null !== $this->retrieveNode('complemento', 'timbreFiscalDigital'));
    }

    /**
     * Normalize a name to follow accesor rules (all is uppercase or first letter is lowercase)
     * - Version => version
     * - TimbreFiscalDigital => timbreFiscalDigital
     * - UUID => UUID
     *
     * @param string $name
     * @return string
     */
    private function normalizeName(string $name): string
    {
        return (strtoupper($name) === $name) ? $name : lcfirst($name);
    }

    /**
     * Utility function to create a child
     *
     * @param SimpleXMLElement $source
     * @param SimpleXMLElement $parent
     * @param array $nss
     * @return SimpleXMLElement
     */
    private function appendChild(SimpleXMLElement $source, SimpleXMLElement $parent, array $nss): SimpleXMLElement
    {
        $new = $parent->addChild($this->normalizeName($source->getName()), (string) $source);
        $this->populateNode($source, $new, $nss);
        return $new;
    }

    /**
     * Utility function to copy contents from one element to other without namespaces
     *
     * @param SimpleXMLElement $source
     * @param SimpleXMLElement $destination
     * @param array $nss
     * @return void
     */
    private function populateNode(SimpleXMLElement $source, SimpleXMLElement $destination, array $nss)
    {
        // populate attributes
        foreach ($nss as $ns) {
            foreach ($source->attributes($ns) as $attribute) {
                /* @var $attribute SimpleXMLElement */
                $destination->addAttribute($this->normalizeName($attribute->getName()), (string) $attribute);
            }
        }
        // populate children
        foreach ($nss as $ns) {
            foreach ($source->children($ns) as $child) {
                $this->appendChild($child, $destination, $nss);
            }
        }
    }

    /**
     * Private helper to locate a node
     * @see node(), attribute()
     * @param string[] ...$nodePath
     * @return null|SimpleXMLElement
     */
    private function retrieveNode(string ...$nodePath)
    {
        $node = $this->comprobante;
        foreach ($nodePath as $level) {
            if (! isset($node->{$level})) {
                return null;
            }
            $node = $node->{$level};
        }
        return $node;
    }
}
